File: Dynamic_Programming/714.Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee/C++/714.Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee.cpp
class Solution {
public:
    int maxProfit(vector<int> & prices, int fee) 
    {
        //======================================//
        // Input type:                          //
        //   - vector<int> ref. (prices vector) //
        //   - int (fee-value)                  //
        // Return type:                         //
        //   - int (result maximun profit)      //
        //======================================//

        /*Initialize*/
        ///// Length of prices vector /////
        int len_prices = prices.size();

        ///// Record dynamic-programming vectors (i.e. earned, cost) /////
        vector<int> record_dp_earned(len_prices, 0), record_dp_cost(len_prices, 0);


        /*Dynamic-programming based loop traversal*/
        record_dp_cost[0] = ((-1) * prices[0]); //Update, Overwrite

        for (int prices_idx = 1; (prices_idx < len_prices); (prices_idx++)) //Whole
        {
            record_dp_earned[prices_idx] = max(record_dp_earned[(prices_idx - 1)], 
                                              (prices[prices_idx] + record_dp_cost[(prices_idx - 1)] - fee)); //Keep updating, overwriting

            record_dp_cost[prices_idx] = max(record_dp_cost[(prices_idx - 1)], (record_dp_earned[prices_idx] - 
                                             prices[prices_idx])); //Keep updating, overwriting

        } //Whole

        return record_dp_earned.back();
    }
};

File: Dynamic_Programming/714.Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee/Python/714.Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee.py
class Solution(object):
    def maxProfit(self, prices, fee):
        """
        :type prices: List[int]
        :type fee: int
        :rtype: int
        """
        #============================================#
        # Dynamic-programming based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of prices array #####
        len_prices = len(prices)

        ##### Record dynamic-programming arrays (i.e. earned, cost) #####
        record_dp_earned, record_dp_cost = [0 for _ in range(len_prices)], [0 for _ in range(len_prices)]


        ##########################################
        #Dynamic-programming based loop traversal
        record_dp_cost[0] = ((-1) * prices[0]) #Update/Overwrite

        for prices_idx in range(1, len_prices):
            record_dp_earned[prices_idx] = max(record_dp_earned[(prices_idx - 1)], 
                                               (prices[prices_idx] + record_dp_cost[(prices_idx - 1)] - fee)) #Keep updating/overwriting

            record_dp_cost[prices_idx] = max(record_dp_cost[(prices_idx - 1)], 
                                             record_dp_earned[prices_idx] - prices[prices_idx]) #Keep updating/overwriting

        return record_dp_earned[(-1)]
